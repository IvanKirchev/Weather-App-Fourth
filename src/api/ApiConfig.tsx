export default {
    openweatherUrl: 'https://api.openweathermap.org',
    // TODO don't store secrets in source control
    openWeatherAppID: '3045bd446337c6e9aa7ef0719614442d'
}