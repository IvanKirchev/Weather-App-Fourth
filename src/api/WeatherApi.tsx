import axios from "axios";
import ApiConfig from "./ApiConfig";

interface IForecast {
    city: ICity
    days: IDayForecast[]
    list: any
}

interface IDayForecast {
    id: number
    name: string,
    averageTemp: number
    forecastByHours: IHourForecast[]
    icon: string,
    minTemp: number,
    maxTemp: number,
    weatherDescription: string
}

interface IHourForecast {
    dt: number
    main: IMainForecastData
    weather: IWeatherDescription[]
    wind: IWind
    visibility: number
}

interface IMainForecastData {
        temp: number
        feels_like: number
        temp_min: number
        temp_max: number
        pressure: number
        sea_level: number
        grnd_level: number
        humidity: number
        temp_kf: number
}

interface IWeatherDescription {
    id: number
    main: string
    description: string
    icon: string
}

interface IWind {
    speed: number
    deg: number
    gust: number
}

interface ICity {
    country: string
    id: number
    name: string
    population: number
    sunrise: number
    sunset: number
    timezone: number
}

function getForecastByCity(city: string, callback: (data: IForecast | null, error: unknown | null) => void) {
    const fetchData = async (): Promise<void> => {
        try {
            const response = await axios.get(`${ApiConfig.openweatherUrl}/data/2.5/forecast?q=${city}&appid=${ApiConfig.openWeatherAppID}&units=metric`)
            callback(response.data, null)
        } catch(error: unknown) {
            callback(null, error)
        }
    };

    fetchData();
}

export { getForecastByCity }
export type { IForecast, IDayForecast, IHourForecast, IMainForecastData, IWeatherDescription, IWind, ICity };
