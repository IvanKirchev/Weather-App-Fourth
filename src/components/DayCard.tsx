import { Grid, Card, CardContent, Typography } from "@mui/material"
import { Box } from "@mui/system"
import react, { Dispatch, SetStateAction, useEffect } from "react"
import { IDayForecast, IHourForecast } from "../api/WeatherApi";

interface DayCardProps {
    selected: boolean
    setSelectedDay: Dispatch<SetStateAction<IDayForecast | null>>
    day: IDayForecast
    setSelectedHour: Dispatch<SetStateAction<IHourForecast | null>>
    selectedHourIndex: number
    setSelectedHourIndex: Dispatch<SetStateAction<number>>
}

function DayCard(props: DayCardProps) {


    // useEffect(() => {
    //    // let newSelectedIndex  = props.day.forecastByHours.length > props.selectedHourIndex ? props.selectedHourIndex : props.day.forecastByHours.length -1;
    //     //props.setSelectedHour(props.day.forecastByHours[props.selectedHourIndex])
    //     //props.setSelectedHourIndex(props.day.forecastByHours.length > props.selectedHourIndex ? props.selectedHourIndex : props.day.forecastByHours.length -1);
    // }, [props.day]);

    return (
        <Grid item xs>
            <Box onClick={() => {
                props.setSelectedDay(props.day)
                
                // let newSelectedIndex  = props.day.forecastByHours.length > props.selectedHourIndex ? props.selectedHourIndex : props.day.forecastByHours.length -1;
                // props.setSelectedHourIndex(props.day.forecastByHours.length > props.selectedHourIndex ? props.selectedHourIndex : props.day.forecastByHours.length -1);
            }
                }>
                <Card variant={props.selected ? "outlined" : "elevation"}>
                    <CardContent>
                        <Typography sx={{ fontSize: 14 }} color="text.secondary" gutterBottom>
                            {props.day.name}
                        </Typography>
                        <Typography variant="h5" component="div">
                            {Math.round(props.day.averageTemp * 10) / 10}Â°C
                        </Typography>
                        <Typography sx={{ mb: 1.5 }} color="text.secondary" position="relative" right="25%">
                            <img src={`http://openweathermap.org/img/w/${props.day.icon}.png`} />
                        </Typography>
                        <Typography variant="h5">
                            {props.day.weatherDescription}
                        </Typography>
                    </CardContent>
                </Card> 
                </Box>
        </Grid>
    )
}

export default DayCard;