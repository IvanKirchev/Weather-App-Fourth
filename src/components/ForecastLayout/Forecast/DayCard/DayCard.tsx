import { Box, Grid, Card, CardContent, Typography } from "@mui/material"

import react, { Dispatch, SetStateAction } from "react"
import { IDayForecast, IHourForecast } from "../../../../api/WeatherApi";

interface DayCardProps {
    selected: boolean
    setSelectedDay: Dispatch<SetStateAction<IDayForecast | null>>
    day: IDayForecast
    setSelectedHour: Dispatch<SetStateAction<IHourForecast | null>>
    selectedHourIndex: number
    setSelectedHourIndex: Dispatch<SetStateAction<number>>
}

function DayCard(props: DayCardProps) {

    return (
        <Grid item xs>
            <Box className="day-card-box" onClick={() => {
                const targetDay = props.day
                const newSelectedHourIndex = targetDay.forecastByHours.length > props.selectedHourIndex ? props.selectedHourIndex : targetDay.forecastByHours.length -1;
                const newSelectedHour = targetDay.forecastByHours[newSelectedHourIndex]

                props.setSelectedDay(targetDay)
                props.setSelectedHourIndex(newSelectedHourIndex)
                props.setSelectedHour(newSelectedHour);
            }
                }>
                <Card sx={{ cursor: 'pointer' }} variant={props.selected ? "outlined" : "elevation"}>
                    <CardContent>
                        <Typography sx={{ fontSize: 14 }} color="text.secondary" gutterBottom>
                            {props.day.name}
                        </Typography>
                        <Typography variant="h5" component="div">
                            {Math.round(props.day.averageTemp * 10) / 10}Â°C
                        </Typography>
                        <Typography sx={{ mb: 1.5 }} color="text.secondary" position="relative" right="25%">
                            <img src={`http://openweathermap.org/img/w/${props.day.icon}.png`} />
                        </Typography>
                        <Typography variant="h5">
                            {props.day.weatherDescription}
                        </Typography>
                    </CardContent>
                </Card> 
                </Box>
        </Grid>
    )
}

export default DayCard;