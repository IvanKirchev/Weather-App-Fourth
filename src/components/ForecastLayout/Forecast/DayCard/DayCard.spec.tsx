import React from "react";
import renderer from "react-test-renderer";
import DayCard from "./DayCard";
import enzyme from "enzyme";
import { IDayForecast } from "../../../../api/WeatherApi";

const mockDayForecast: IDayForecast = {
  id: 1637236800000,
  name: "Thursday",
  order: 0,
  icon: "04d",
  forecastByHours: [],
  weatherDescription: "broken clouds",
  averageTemp: 7.445,
  minTemp: 0,
  maxTemp: 0,
};

it("Should render DayCard component correctly", () => {
  const component = renderer.create(
    <DayCard
      day={mockDayForecast}
      selected={false}
      setSelectedDay={() => {}}
      setSelectedHour={() => {}}
      selectedHourIndex={0}
      setSelectedHourIndex={() => {}}
    />
  );
  expect(component).toMatchSnapshot();
});

it("Should call on a card click the following functions: setSelectedDat, setSelectedHour, setSelectedHourIndex", () => {
  const spySetSelectedDay = jest.fn();
  const spySetSelectedHour = jest.fn();
  const spySetSelectedHourIndex = jest.fn();

  const component = enzyme.shallow(
    <DayCard
      day={mockDayForecast}
      selected={false}
      setSelectedDay={spySetSelectedDay}
      setSelectedHour={spySetSelectedHour}
      selectedHourIndex={0}
      setSelectedHourIndex={spySetSelectedHourIndex}
    />
  );

  component.find(".day-card-box").at(0).simulate("click")

  expect(spySetSelectedDay).toHaveBeenCalled();
  expect(spySetSelectedHour).toHaveBeenCalled();
  expect(spySetSelectedHourIndex).toHaveBeenCalled();
});
