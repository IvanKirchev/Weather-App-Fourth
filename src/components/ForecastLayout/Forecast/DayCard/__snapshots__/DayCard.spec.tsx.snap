// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render DayCard component correctly 1`] = `
<div
  className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-4jz4pb-MuiGrid-root"
>
  <div
    className="day-card-box MuiBox-root css-0"
    onClick={[Function]}
  >
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-2v22o2-MuiPaper-root-MuiCard-root"
    >
      <div
        className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-16m3jxf-MuiTypography-root"
        >
          Thursday
        </p>
        <div
          className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          7.4
          Â°C
        </div>
        <p
          className="MuiTypography-root MuiTypography-body1 css-9eg9l8-MuiTypography-root"
        >
          <img
            src="http://openweathermap.org/img/w/04d.png"
          />
        </p>
        <h5
          className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          broken clouds
        </h5>
      </div>
    </div>
  </div>
</div>
`;
